sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverClientIntegrationTest.java

```diff
@@ -0,0 +1,595 @@
+     * Receiver used to clean up resources in {@link #afterTest()}.
+     */
+    private ServiceBusReceiverClient receiveAndDeleteReceiver;
+
+    protected ServiceBusReceiverClientIntegrationTest() {
+        super(new ClientLogger(ServiceBusReceiverClientIntegrationTest.class));
+    }
+
+    @Override
+    protected void beforeTest() {
+        sessionId = UUID.randomUUID().toString();
+    }
+
+    @Override
+    protected void afterTest() {
+        final int pending = messagesPending.get();
+        if (pending < 1 && messagesDeferred.get().size() < 1) {
+            dispose(receiver, sender, receiveAndDeleteReceiver);
+            return;
+        }
+        // In the case that this test failed... we're going to drain the queue or subscription.
+        if (pending > 0) {
+            try {
+                IterableStream<ServiceBusReceivedMessage> removedMessage = receiveAndDeleteReceiver.receive(
+                    pending + BUFFER_MESSAGES_TO_REMOVE);
+                removedMessage.stream().forEach(receivedMessage -> {
+                    logger.info("Removed Message Seq: {} ", receivedMessage.getSequenceNumber());
+                });
+            } catch (Exception e) {
+                logger.warning("Error occurred when draining queue.", e);
+            }
+        }
+
+        if (messagesDeferred.get().size() > 0) {
+            try {
+                List<Long> listOfDeferredMessages = messagesDeferred.get();
+                for (Long seqNumber : listOfDeferredMessages) {
+                    ServiceBusReceivedMessage deferredMessages = receiver.receiveDeferredMessage(seqNumber);
+                    receiver.complete(deferredMessages);
+                }
+            } catch (Exception e) {
+                logger.warning("Error occurred when draining deferred messages Entity: {} ", receiver.getEntityPath(), e);
+            }
+
+            // set empty list
+            messagesDeferred.set(new ArrayList<>());
+        }
+
+        dispose(receiver, sender, receiveAndDeleteReceiver);
+    }
+
+    /**
+     * Verifies that we can send and receive two messages.
+     */
+    @MethodSource("messagingEntityWithSessions")
+    @ParameterizedTest
+    void receiveTwoMessagesAndComplete(MessagingEntityType entityType, boolean isSessionEnabled) {
+        // Arrange
+        setSenderAndReceiver(entityType, isSessionEnabled);
+        int maxMessages = 2;
+
+        final String messageId = UUID.randomUUID().toString();
+        final ServiceBusMessage message = getMessage(messageId, isSessionEnabled);
+
+        sendMessage(message);
+        sendMessage(message);
+
+        // Act
+        IterableStream<ServiceBusReceivedMessage> messages = receiver.receive(maxMessages, TIMEOUT);
+
+        // Assert
+        final List<ServiceBusReceivedMessage> asList = messages.stream().collect(Collectors.toList());
+        assertEquals(maxMessages, asList.size());
+
+        for (ServiceBusReceivedMessage receivedMessage : asList) {
```

Comment:
A better way is to count the number of time you go into the loop and assert at the end rather than creating the list eagerly.

```java
int count = 0;
for (Message m : messages) {
    count++;
}
assertEquals(2, count);
```

Reply:
Same with other instances.